src/util/std_expr.h

```diff
@@ -3094,30 +3086,26 @@ inline void validate_expr(const not_exprt &value)
 
 /*! \brief Operator to dereference a pointer
 */
-class dereference_exprt:public exprt
+class dereference_exprt:public unary_exprt
 {
 public:
-  explicit dereference_exprt(const typet &type):
-    exprt(ID_dereference, type)
+  dereference_exprt():unary_exprt(ID_dereference)
   {
-    operands().resize(1);
   }
 
-  explicit dereference_exprt(const exprt &op):
-    exprt(ID_dereference)
+  explicit dereference_exprt(const typet &type):
+    unary_exprt(ID_dereference, type)
   {
-    copy_to_operands(op);
   }
 
-  dereference_exprt(const exprt &op, const typet &type):
-    exprt(ID_dereference, type)
+  explicit dereference_exprt(const exprt &op):
+    unary_exprt(ID_dereference, op)
```

Comment:
This would yield the wrong type, it should be `op.subtype()` if `op.type().id()` is ID_pointer (and all other cases are invalid)

Reply:
Indeed, fixed. Note that I've added a precondition, which is a new thing.